VM to VM calls

Schaeff opened the pull request:
This PR introduces a more modular compilation pipeline, with the following steps:
- macro_expansion (unchanged)
- type checking (unchanged)
- start asm analysis: any machine with no pc is unaffected by the following steps
  - infer assignment registers (unchanged)
  - batch asm statements (unchanged, but now happens for each function instead of after romgen)
  - One bigger step which does the following:
    - Generate the ROM
      - Introduce input and output registers common to all functions
      - Substitute them in the functions
      - Inline the functions into the ROM as labels. This introduces a few internal instructions.
    - Reduce the ROM to pil (almost unchanged, except unconstraining registers on the first line as constraining them is the responsibility of the dispatcher and some other things we need to change for the dispatcher to work, which ideally would go away)
- end asm analysis. from now on, everything is a constrained machine!
- enforce blocks, which means that `function_id` can only change when `latch` is 1
- ~generate air, which is now trivial: set degree, get all pil constraints for each machine, generate links for all external instructions~ extracted to #505 

Status:
- [x] fails in witness generation: will be fixed in another PR wink wink @georgwiese 
- [x] assumes the input registers are not written to, they should be made properly read-only
- [x] still some unclarity around inputs and outputs, to be clarified
- [x] fix tests

Schaeff commented on asm_to_pil/src/lib.rs:
```diff
@@ -215,6 +111,13 @@ impl<'a, T: FieldElement> ASMPILConverter<'a, T> {
                             )
                         }
+                        // Unconstrain the inputs when calling `_reset`
+                        if name.starts_with("_input") {
```
hacky

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
     pub fn compile_simple_sum() {
         let expectation = r#"
 namespace main(1024);
+pol commit XInv;
+pol commit XIsZero;
+XIsZero = (1 - (X * XInv));
+(XIsZero * X) = 0;
+(XIsZero * (1 - XIsZero)) = 0;
```
reordered

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
+XIsZero = (1 - (X * XInv));
+(XIsZero * X) = 0;
+(XIsZero * (1 - XIsZero)) = 0;
+pol commit _sigma;
+_sigma' = ((1 - first_step') * (_sigma + instr_return));
+(_sigma * (_function_id - 10)) = 0;
```
Force witgen to hit the infinite loop after the first block, will be removed

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
 pol constant first_step = [1] + [0]*;
 pol commit pc;
 pol commit X;
-(first_step * A) = 0;
```
register reset happens on `instr__reset`

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
-(first_step * A) = 0;
 pol commit reg_write_X_A;
 pol commit A;
-(first_step * CNT) = 0;
```
same

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
-(first_step * CNT) = 0;
 pol commit reg_write_X_CNT;
 pol commit CNT;
-pol commit XInv;
-pol commit XIsZero;
-XIsZero = (1 - (X * XInv));
-(XIsZero * X) = 0;
-(XIsZero * (1 - XIsZero)) = 0;
```
reordered

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
 pol commit instr_dec_CNT;
 pol commit instr_assert_zero;
 (instr_assert_zero * (XIsZero - 1)) = 0;
+pol commit instr_return;
```
return instruction exposed to the user

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
 pol commit instr_assert_zero;
 (instr_assert_zero * (XIsZero - 1)) = 0;
+pol commit instr_return;
+pol commit instr__jump_to_operation;
+pol commit instr__reset;
+pol commit instr__loop;
```
dispatcher instructions hidden from the user

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
+pol constant p_reg_write_X_A = [0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0] + [0]*;
+pol constant p_reg_write_X_CNT = [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0] + [0]*;
+{ pc, reg_write_X_A, reg_write_X_CNT, instr_jmpz, instr_jmpz_param_l, instr_jmp, instr_jmp_param_l, instr_dec_CNT, instr_assert_zero, instr_return, instr__jump_to_operation, instr__reset, instr__loop, X_const, X_read_free, read_X_A, read_X_CNT, read_X_pc } in { p_line, p_reg_write_X_A, p_reg_write_X_CNT, p_instr_jmpz, p_instr_jmpz_param_l, p_instr_jmp, p_instr_jmp_param_l, p_instr_dec_CNT, p_inst ...
+pol constant _first_step = [1] + [0]*;
```
Introduced by the block enforcer, named differently from `first_step` which is introduced by asm_to_pil to keep things independent, could be optimised away

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
+pol constant p_reg_write_X_CNT = [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0] + [0]*;
+{ pc, reg_write_X_A, reg_write_X_CNT, instr_jmpz, instr_jmpz_param_l, instr_jmp, instr_jmp_param_l, instr_dec_CNT, instr_assert_zero, instr_return, instr__jump_to_operation, instr__reset, instr__loop, X_const, X_read_free, read_X_A, read_X_CNT, read_X_pc } in { p_line, p_reg_write_X_A, p_reg_write_X_CNT, p_instr_jmpz, p_instr_jmpz_param_l, p_instr_jmp, p_instr_jmp_param_l, p_instr_dec_CNT, p_inst ...
+pol constant _first_step = [1] + [0]*;
+pol commit _function_id;
+pol commit _function_id_no_change;
+_function_id_no_change = ((1 - _first_step') * (1 - instr_return));
+(_function_id_no_change * (_function_id' - _function_id)) = 0;
```
block enforcement

Schaeff commented on linker/src/lib.rs:
```diff
@@ -185,52 +204,67 @@ mod test {
+pol commit _function_id_no_change;
+_function_id_no_change = ((1 - _first_step') * (1 - instr_return));
+(_function_id_no_change * (_function_id' - _function_id)) = 0;
+(first_step * (_function_id - 2)) = 0;
```
Force the main machine to start with a block of `main`

leonardoalt commented on riscv/tests/.DS_Store:
wat

leonardoalt commented on airgen/src/lib.rs:
```diff
@@ -0,0 +1,193 @@
+//! Compilation from powdr machines to AIRs
+
```
Was this file just moved from somewhere or are there new things here?

Schaeff replied:
TL;DR: `airgen` is the extraction of the part of `asm_to_pil` which is not ASM specific. So it's `asm_to_pil` moved here and stripped down.

Before:
`asm_to_pil` turns each machine into an AIR, reducing the ASM to constraints.

After:
`asm_to_pil` only turns each machine into a another machine where all the ASM stuff (rom, instructions*, registers) is turned into constraints.

`airgen` turns each machine into an AIR

The reasoning is that after `asm_to_pil`, all machines are block machines, so we can do more processing steps before hitting `airgen`

*external instructions are still kept, so we still have some ASM information after `asm_to_pil`. This is fixed in #464 where external instructions are turned into links, which are the low-level type of machine-to-machine interaction.

leonardoalt replied:
I see. Should `asm_to_pil` be renamed then? Or is the output still an `Analyzed` PIL?

Schaeff replied:
The output is the same as the input. Maybe `asm_encoder` ?

leonardoalt commented on ast/src/parsed/asm.rs:
```diff
@@ -91,6 +91,7 @@ pub enum DebugDirective {
 pub enum RegisterFlag {
     IsPC,
     IsAssignment,
+    IsReadOnly,
```
What's a readonly register?

Schaeff replied:
https://github.com/powdr-labs/powdr/blob/call-to-vm/book/src/asm/registers.md#read-only-registers

leonardoalt replied:
I don't quite get why they're called `readonly`. If they're used for inputs, they're also sort of "written" from the caller machine?

Schaeff replied:
Right, they are read-only within a block: they get set non-deterministically at the beginning of each block, and then they cannot change within this block. Any better name? `FunctionInput` ?

leonardoalt commented on ast/src/object/display.rs:
```diff
@@ -56,11 +56,17 @@ impl Display for Instr {
             f,
-            "// {} in object with latch \"{}\" and function_id \"{}\"",
-            self.function, self.latch, self.function_id
+            "object with latch \"{}\" and function_id \"{}\"",
```
```suggestion
            "// object with latch \"{}\" and function_id \"{}\"",
```
?

Schaeff replied:
No, it's the second part of the `LinkTo` format

leonardoalt replied:
sorry, I'm still a bit lost. Does this go in a comment line or is this actual code?

Schaeff replied:
it goes in the second part of a comment line for LinkTo
// operation foo in **object with latch ...**

leonardoalt replied:
ok!

leonardoalt commented on test_data/asm/vm_to_vm.asm:
```diff
@@ -0,0 +1,45 @@
+
+    instr sub X, Y -> Z { X - Y = Z }
+
+    function add x: field, y: field -> field {
```
do we really need the types already?

Schaeff replied:
Good call. So the types are not checked anywhere, but I wasn't sure how to express the signature without them. Any ideas?

leonardoalt replied:
maybe same as instr?

My main point here is that if we don't need the types it simplifies the PR. There are quite a few new features already, it would be good to have only the essential ones.

Schaeff replied:
But for `instr` you have registers. I could change to `function add x, y -> _, _`? But it wouldn't change any of the code, because `field` is not checked anywhere.

leonardoalt replied:
ok

leonardoalt reviewed (commented):
some comments about tests and syntax

leonardoalt commented on riscv/src/compiler.rs:
```diff
@@ -1358,6 +1394,11 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
         }
 
+        // The powdr embedded `return` instruction
+        "return" => {
```
I don't think RISCV has an instruction called `return`

Schaeff replied:
Right.. I added it so that when I replace
```
tail main;
```
by
```
call main;
return;
```

The `return` gets turned into powdr asm. I guess I should instead run this substitution on the powdr asm.

leonardoalt replied:
yea I think so too

Schaeff replied:
moved to the asm to powdr asm match

leonardoalt commented on linker/src/lib.rs:
```diff
@@ -82,27 +83,45 @@ pub fn link<T: FieldElement>(graph: PILGraph<T>) -> Result<PILFile<T>, Vec<Strin
+                if let Some(entry_point) = graph.entry_points.first() {
+                    let entry_point_id = entry_point.id;
+                    let function_id = main_machine.function_id.clone();
+                    // call the first function by initialising `function_id` to the first function
```
but first is the single function?

leonardoalt replied:
it's asserted above that there is only one function, I just thought the comment could be clearer

Schaeff replied:
I see will fix

Schaeff replied:
Called main instead

leonardoalt commented on linker/src/lib.rs:
```diff
@@ -185,52 +209,67 @@ mod test {
 pol commit instr_assert_zero;
 (instr_assert_zero * (XIsZero - 1)) = 0;
+pol commit instr_return;
+pol commit instr__jump_to_operation;
+pol commit instr__reset;
+pol commit instr__loop;
```
why the double underscore

Schaeff replied:
Because they are the flags for the instructions `_jump_to_operation` `_reset` `_loop`
This is my way of indicating that they are not user facing.

leonardoalt replied:
ok

leonardoalt commented on compiler/tests/asm.rs:
```diff
@@ -102,12 +111,14 @@ fn full_pil_constant() {
-
-        verify_asm::<GoldilocksField>(f.to_str().unwrap(), slice_to_vec(&i));
-        gen_halo2_proof(f.to_str().unwrap(), slice_to_vec(&i));
+        if [".asm", ".pil"].contains(&f.extension().unwrap().to_str().unwrap()) {
```
Ah I see, `.ignore` suffix is for this

leonardoalt commented:
> generate air, which is now trivial: set degree, get all pil constraints for each machine, generate links for all external instructions

That's a mathematicians notion of trivial :)

Schaeff commented on test_data/asm/vm_to_block_unique_interface.asm:
```diff
@@ -59,17 +59,13 @@ machine Main {
+        A <== or(A, 0);
         assert_eq A, 1;
-        loop;
+        return;
```
There is no implicit return now.
In the absence of a return statement, we currently have undefined behavior as the dispatcher does not prevent falling through after the end of the body. I was thinking of having a separate PR to add this protection, in which case no return would panic. I would be in favor of implicit returns as well though but that's a separate question.

leonardoalt replied:
ok, fine with UB in this PR, but we should keep track of that and probably implement the implicit behavior (whichever it is) right after

Schaeff replied:
https://github.com/powdr-labs/powdr/issues/482

Schaeff commented on test_data/asm/book/asm_function_after.asm.ignore:
```diff
@@ -0,0 +1,12 @@
+    function foo _input_0, _input_1 -> _output_0 {
+        return (_input_0 + _input_1);
+    }
+}
```
This is what a function gets desugared into. So yes `x: field` ends up being substituted for `reg[@r] _input_0`.
One thing we could do to make it less magic (with the global `_input_0` registers) is creating input registers for each function:
```
reg _foo_x[@r];
reg _bar_x[@r];

function foo _foo_x {
   ...
}

function bar _bar_x {
   ...
}
```
And optimise them away later

leonardoalt replied:
I need to think more about this myself, I don't understand what's happening

Schaeff replied:
I suggest you run `RUST_LOG=trace cargo run pil test_data/asm/vm_to_vm.asm` and look at the diffs, can also explain on a call

Schaeff commented on linker/src/lib.rs:
```diff
@@ -82,27 +83,45 @@ pub fn link<T: FieldElement>(graph: PILGraph<T>) -> Result<PILFile<T>, Vec<Strin
+                    let function_id = main_machine.function_id.clone();
+                    // call the first function by initialising `function_id` to the first function
+                    pil.push(parse_pil_statement(&format!(
+                        "first_step * ({function_id} - {entry_point_id}) = 0"
```
What I can do is redeclaring a new `first_step` for the linker step, so that it's all self contained. Then all the `first_step` can be optimised away.

leonardoalt replied:
yea I don't have a solution for it myself, but self-contained sounds good. What I'd like to avoid is assuming that something called `first_step` exists here, although currently there are no connections to it's declaration. So if that changes wherever it is now, this breaks in a separate step

Schaeff replied:
Totally agree, self contained is the way then

Schaeff commented on linker/src/lib.rs:
```diff
@@ -132,14 +151,19 @@ mod test {
         // a graph with two objects of degree `main_degree` and `foo_degree`
         let test_graph = |main_degree, foo_degree| PILGraph {
-            main: Location::from("main".to_string()),
+            main: ast::object::Machine {
+                location: Location::from("main".to_string()),
+                function_id: "function_id".into(),
+                latch: "latch".into(),
```
This is a test though, these values do not matter here

leonardoalt replied:
ah true

Schaeff commented on compiler/tests/asm.rs:
```diff
@@ -60,6 +60,15 @@ fn vm_to_block_multiple_interfaces() {
 }
 
+#[test]
+#[should_panic = "not implemented: No executor machine matched identity `main.instr_add { 2, main.X, main.Y, main.Z } in main_vm.instr_return { main_vm._function_id, main_vm._input_0, main_vm._input_1, main_vm._output_0 };`"]
```
This is witgen not knowing what to do about a call to a VM.

leonardoalt replied:
ah I see. Is this gonna be fixed in this PR?

georgwiese replied:
Nope, the plan is to fix it later.

leonardoalt replied:
ok

georgwiese reviewed (commented):
Very cool stuff! 🎉 

I read through most of the code at least briefly and followed the steps on the `vm_to_vm` example. The diffs are dope!

Schaeff commented on book/src/asm/registers.md:
```diff
@@ -37,4 +37,21 @@ For example, if we want to assert that write register `A` is `0`, we can use the
+
+// After:
+
+{{#include ../../../test_data/asm/book/asm_function_after.asm}}
```
Good call! Also noticed it on another branch. Should add an URI checker to CI.

Schaeff replied:
updated URI

Schaeff commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+
+    fn generate_machine_rom(&self, mut machine: Machine<T>) -> Machine<T> {
+        if machine.has_pc() {
+            let latch = "instr_return";
```
Right! So the only user-facing one is `return`, so it could be banned in the parser or type checker (any preference @leonardoalt?). The other ones start with `_` which the user is not expected to use. This could be enforced as well.

leonardoalt replied:
yea fine with either parser or type checker

Schaeff replied:
Added to type checker

Schaeff commented on test_data/asm/vm_to_vm.asm:
```diff
@@ -0,0 +1,45 @@
+    // this should be the identity function: x + 1 - 1 = x
+    function main {
+        A <=Z= add(1, 1);
+        A <=Z= sub(A, 1);
```
Yes, I'll comment out and create an issue with everything that needs to happen once witgen works.

Schaeff replied:
Commented out and added todo, issue here #487

Schaeff commented on test_data/asm/vm_to_vm.asm:
```diff
@@ -0,0 +1,45 @@
+    function main {
+        A <=Z= add(1, 1);
+        A <=Z= sub(A, 1);
+        assert_eq A, 1;
```
True, return statement needed

Schaeff replied:
Added return

georgwiese replied:
I think we should either have implicit returns or fail if the last instruction isn't return.

Schaeff commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+                        assert_eq!(
+                            outputs.params.len(),
+                            1,
+                            "number of function outputs should be 0 or 1"
```
I'll look into this, maybe we don't need this restriction.

Schaeff replied:
Now multiple outputs are allowed. It happens in romgen:
- romgen finds the max number of outputs
- all return statements are padded with zeroes to that max number
- when we return, we set the output regs to `[a, b, c, 0, ..., 0]`. From the outside we only access `[a, b, c]` as we rely on the original signature of the function

leonardoalt commented on analysis/src/block_enforcer.rs:
```diff
@@ -0,0 +1,37 @@
+    for machine in file.machines.values_mut() {
+        let function_id = machine.function_id.as_ref().unwrap();
+        let latch = machine.latch.as_ref().unwrap();
+        let first_step = "_first_step";
```
so every machine has their own first step column?

Schaeff replied:
Yes! They are different AIRs in the end.

leonardoalt commented on analysis/src/block_enforcer.rs:
```diff
@@ -0,0 +1,37 @@
+        // add the necessary embedded constraints which apply to both static and dynamic machines
+        let embedded_constraints = [
+            // inject `_first_step`
+            parse_pil_statement(&format!("pol constant {first_step} = [1] + [0]*")),
```
pol -> col, also below

Schaeff replied:
done

leonardoalt commented on analysis/src/block_enforcer.rs:
```diff
@@ -0,0 +1,37 @@
+            // inject the function_id
+            parse_pil_statement(&format!("pol witness {function_id}")),
+            // the function id must be constant within a block
+            parse_pil_statement(&format!("pol witness {f_no_change}")),
```
this could also be an intermediate?

Schaeff replied:
Good catch

leonardoalt commented on analysis/src/block_enforcer.rs:
```diff
@@ -0,0 +1,37 @@
+            // the function id must be constant within a block
+            parse_pil_statement(&format!("pol witness {f_no_change}")),
+            parse_pil_statement(&format!(
+                "{f_no_change} = (1 - {first_step}') * (1 - {latch})"
```
is this relying on a cyclic constraint? I'm asking because `first_step` only appears in constraints as `next` here

leonardoalt commented on analysis/src/vm/batcher.rs:
```diff
@@ -23,10 +25,13 @@ impl<'a, T: FieldElement> Batch<'a, T> {
     }
 
     /// Returns true iff this batch consists exclusively of labels
```
`and directives`?

Schaeff replied:
changed

leonardoalt commented on analysis/src/vm/batcher.rs:
```diff
@@ -135,31 +145,50 @@ mod tests {
 
-    use number::GoldilocksField;
+    use ast::asm_analysis::AnalysisASMFile;
+    use number::Bn254Field;
```
why the field change?

Schaeff replied:
doesn't matter, reverted

leonardoalt commented on analysis/src/vm/batcher.rs:
```diff
@@ -135,31 +145,50 @@ mod tests {
+        let batched_str = batched.to_string();
+
+        assert!(batched_str.contains("instr return {  }"));
+        assert!(batched_str.contains("(instr_return, _)"));
```
what's this one?

Schaeff replied:
added comment

leonardoalt commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+
+#[derive(Default)]
+struct FunctionDesugar<T> {
+    marker: PhantomData<T>,
```
still find this super silly from Rust

Schaeff replied:
removed the struct entirely as it has no state

leonardoalt commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+fn substitute<T>(mut e: Expression<T>, inputs: &HashMap<String, usize>) -> Expression<T> {
+    previsit_expression_mut(&mut e, &mut |e| {
```
is there a version of this but for statements that also applies it to expressions? if yes we could maybe avoid the big replacement match lower in this file

leonardoalt commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+            machines: file
+                .machines
+                .into_iter()
+                .map(|(name, m)| (name, self.generate_machine_rom(m)))
```
why is this file generating the rom? that feels like more than desugaring?

Schaeff replied:
oops wrong name from when I extracted this! it does not generate the rom. will update the name

leonardoalt commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+            machine
+                .instructions
+                .push(parse_instruction_definition(&format!(
+                    "instr return {} {{ }}",
```
so this is parsed now but only filled later?

Schaeff replied:
Right, it's up to the dispatcher how `return` is implemented, but at this point we already know that the latch will be `instr_return`

leonardoalt commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+                        .params
+                        .iter()
+                        .enumerate()
+                        .map(|(index, param)| (param.name.clone(), index))
```
so this bit and line 96 above rely on the same enumeration but they're not connected. could they be done together somehow, or chained directly? also applicable to line 122 below

Schaeff replied:
I don't think so: declaration of the input and output registers is global to the machine, while remapping of function inputs and outputs to these global inputs and outputs happens at the function level.

leonardoalt commented on analysis/src/vm/function_desugar.rs:
```diff
@@ -0,0 +1,207 @@
+            std::fs::read_to_string("../test_data/asm/book/asm_function_before.asm").unwrap();
+        let desugared: AnalysisASMFile<Bn254Field> = function_desugar_str(&before);
+        // TODO: this file does not quite compile yet because the latch `instr_return` conflicts with the existence of a pc.
+        // this should be fixed so that this file can be compiled
```
still the case?

Schaeff replied:
Yes this is still the case.

leonardoalt commented on analysis/src/vm/mod.rs:
```diff
@@ -0,0 +1,73 @@
+//! Analysis for VM machines, reducing them to PIL
```
the return type is AnalysisASMFile, maybe the comment needs to be updated?

Schaeff replied:
changed

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+};
+
+/// Generate the ROM for each machine based on its functions
+pub fn generate_rom<T: FieldElement>(file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
```
I think it would be good to have an entry in the book or somewhere (maybe a README here) about this process. It's doing quite a lot of things that are hard to get without a previous explanation just from the code.

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )),
+            parse_instruction_definition(&format!("instr _loop {{ {}' = {} }}", pc, pc)),
```
{pc} {pc}

Schaeff replied:
done

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+
+        machine.instructions.extend(embedded_instructions);
+
+        // implement the return instruction. For this dispatcher, returning jumps to 0
```
why 0?

Schaeff replied:
it's where we enter in the first place. we go back there after each block

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+        machine
+            .instructions
+            .iter_mut()
+            .find(|i| i.name == "return")
```
this feels like it should be an enum

Schaeff replied:
That would go in the direction of making `return` a special keyword, not a normal instruction. I don't necessarily disagree, but this could be done in a later PR?

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+                .reason(IncompatibleSet::from(Incompatible::Label)),
+        ]);
+
+        // add each function, setting the function_id to the current position in the ROM
```
where is this happening? I don't see `function_id` being used in the loop

Schaeff replied:
```rust
function.id = Some(T::from(rom.len() as u64));
```

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+            rom.extend(batches);
+        }
+
+        // THE FOLLOWING IS NECESSARY BECAUSE OF WITGEN, IT COULD BE REMOVED ONCE WITGEN CAN CALL THE INFINITE LOOP ITSELF
```
might make sense to add `TODO:` in front and open an issue

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+
+        // THE FOLLOWING IS NECESSARY BECAUSE OF WITGEN, IT COULD BE REMOVED ONCE WITGEN CAN CALL THE INFINITE LOOP ITSELF
+        // we get the location of the sink so that witgen jumps to it after the first call is done
+        // this constrains the VM to being able to execute only one call, which will be fixed in the future
```
might make sense to add `TODO:` in front and open an issue

Schaeff replied:
added TODO and #487

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+        // THE FOLLOWING IS NECESSARY BECAUSE OF WITGEN, IT COULD BE REMOVED ONCE WITGEN CAN CALL THE INFINITE LOOP ITSELF
+        // we get the location of the sink so that witgen jumps to it after the first call is done
+        // this constrains the VM to being able to execute only one call, which will be fixed in the future
+        let sink_id = T::from(rom.len() as u64);
```
this value is also used by the loop above, could be declared before

Schaeff replied:
No, this value changes as we add statements to the rom. Related to your question above.

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+                // declare `_sigma` as the sum of the latch, will be 0 and then 1 after the end of the first call
+                parse_pil_statement(&format!("col witness {sigma}")),
+                parse_pil_statement(&format!(
+                    "{sigma}' = (1 - first_step') * ({sigma} + {latch})"
```
is this cyclic? (because of first_step')

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+                parse_pil_statement(&format!("col witness {sigma}")),
+                parse_pil_statement(&format!(
+                    "{sigma}' = (1 - first_step') * ({sigma} + {latch})"
+                )), // HACKY: THIS ASSUMES `first_step` is defined here!!
```
what's the fix?

Schaeff replied:
made it self contained!

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+            
+                instr sub X, Y -> Z { X - Y = Z }
+            
+                function add x: field, y: field -> field {
```
I don't think function and instr names should be allowed to collide, it can look confusing

Schaeff replied:
Fair enough. Changed the source here. Added #489

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+                instr sub X, Y -> Z { X - Y = Z }
+            
+                function add x: field, y: field -> field {
+                    A <=Z= add(x, y);
```
<==

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+                }
+            
+                function sub x: field, y: field -> field {
+                    A <=Z= sub(x, y);
```
<==

leonardoalt commented on analysis/src/vm/romgen.rs:
```diff
@@ -0,0 +1,217 @@
+"#
+            .replace('\t', "    ")
+            .trim()
+        );
```
It would be nice if the test also checked the generated instructions.

leonardoalt commented on asm_to_pil/src/lib.rs:
```diff
@@ -122,43 +56,19 @@ struct ASMPILConverter<'a, T> {
-        self.submachines = input.submachines;
-
+impl<T: FieldElement> ASMPILConverter<T> {
+    fn convert_machine(mut self, mut input: Machine<T>) -> Machine<T> {
```
The signature of this function changed. What does it convert a Machine into?

Schaeff replied:
It turns a virtual machine into a constrained machine. It used to go all the way to AIR, but now there are more processing steps before we go to air in airgen.

leonardoalt commented on asm_to_pil/src/lib.rs:
```diff
@@ -206,21 +105,31 @@ impl<'a, T: FieldElement> ASMPILConverter<'a, T> {
+                            Pc => (
+                                lhs,
+                                build_mul(
+                                    build_sub(build_number(1u64), next_reference("first_step")),
```
so `first_step` also shows up here, completely disconnected from the internal machine to compiler which is in a different crate. I think we need to address this now somehow before we get too far

Schaeff replied:
Agreed. I used other names in other compiler steps. I kept this name as this is what it was called before this PR. I could also rename to `_asm_to_pil_first_step`.

leonardoalt commented on asm_to_pil/src/lib.rs:
```diff
@@ -206,21 +105,31 @@ impl<'a, T: FieldElement> ASMPILConverter<'a, T> {
+                            // Unconstrain read-only registers when calling `_reset`
+                            ReadOnly => {
+                                let not_reset =
+                                    build_sub(build_number(1u64), direct_reference("instr__reset"));
```
same thing here, using a hardcoded constant that is defined in a different crate

Schaeff replied:
This is fine, as `first_step` is declared here. I changed the others.

leonardoalt commented on asm_to_pil/src/lib.rs:
```diff
@@ -352,32 +263,24 @@ impl<'a, T: FieldElement> ASMPILConverter<'a, T> {
+        s: InstructionDefinitionStatement<T>,
+    ) -> Option<InstructionDefinitionStatement<T>> {
+        let instruction_name = s.name.clone();
+        let instruction_flag = format!("instr_{instruction_name}");
```
same here

Schaeff replied:
Agreed. This compiler steps assumes the existence of `instr_return` and `instr__reset`, which relies on the existence of `return` and `reset`. I do not have an immediate solution to that, apart from grouping all compiler steps together, which is not ideal. I'll have more of a think about it.

Schaeff replied:
Decided to make `return` first class citizen in functions and `return` and `reset` first class citizens in rom.

Schaeff replied:
Implemented only the former for now

leonardoalt commented on ast/src/asm_analysis/display.rs:
```diff
@@ -1,77 +1,81 @@
-            }
-        }
-        write!(f, "\t// }}")
+        writeln!(f, "rom {{")?;
```
This was commented out before, why not anymore?

Schaeff replied:
It just moved to the caller
```rust
writeln!(f, "{}", indent(comment_out(rom), 1),)?;
```

leonardoalt reviewed (commented):
Lots of comments, but most are small.
I think the PR probably looks good overall, but my main concerns right now are:

- [ ] There are a lot of new things going on here. New features, new concepts, multiple changes. I think it would be nice to have a README somewhere (maybe VM) or book entry explaining how the new translation works, with a full example. I know you @Schaeff and @georgwiese talked about this in person, and I do have the impression that each new feature/change is small and easy to understand, but I found the combo of everything together a bit hard to unpack just from the code and the given tests/examples.
- [ ] There are some hardcoded column and registers names being used in 2 different crates without a programmatic construction between them. I think this shouldn't be the case.
- [ ] I think some more tests would be nice.

Schaeff commented on test_data/asm/vm_to_vm.asm:
```diff
@@ -0,0 +1,45 @@
+    instr sub X, Y -> Z = vm.sub
+    instr assert_eq X, Y { X = Y }
+
+    // this should be the identity function: x + 1 - 1 = x
```
updated the code

Schaeff commented on test_data/asm/vm_to_block_unique_interface.asm:
```diff
@@ -1,6 +1,6 @@
 machine Binary(latch, function_id) {
 
-    degree 8;
+    degree 32;
```
removed

Schaeff commented on test_data/asm/vm_to_vm.asm:
```diff
@@ -0,0 +1,45 @@
+
+    // this should be the identity function: x + 1 - 1 = x
+    function main {
+        A <=Z= add(1, 1);
```
removed

Schaeff commented on airgen/src/lib.rs:
```diff
@@ -0,0 +1,193 @@
+
+use number::FieldElement;
+
+pub fn compile<T: FieldElement>(input: AnalysisASMFile<T>) -> PILGraph<T> {
```
Done

Schaeff commented:
Update:
- [ ] This is blocked by a witgen issue with block machines, @georgwiese is on it
- [x] I added a README in `analysis`
- [x] I removed hardcoded stuff and put it in `common.rs` instead
- [x] I added romgen tests. @leonardoalt any other specific tests I should add?

Schaeff commented on asm_to_pil/src/vm_to_constrained.rs:
```diff
@@ -0,0 +1,1001 @@
+//! Compilation from powdr assembly to PIL
```
Moved from `asm_to_pil/src/lib.rs` and modified

georgwiese reviewed (commented):
Very cool! I went through the compiled PILs again, looks good to me! I also skimmed the code, looks very readable and understandable to me. I also verified that #503 indeed fixes the failing test.

Schaeff commented on test_data/asm/poseidon_bn254.asm:
```diff
@@ -145,6 +145,6 @@ machine Main {
         assert_eq A, 0x188ada144ed909426b0396e967a82e26d739652cff288d13306279d91f29010c;
 
-        loop;
+        return;
```
Yeah whichever comes first. We can fix this once we have imports.

Schaeff commented on analysis/src/block_enforcer.rs:
```diff
@@ -0,0 +1,38 @@
+
+use crate::utils::parse_pil_statement;
+
+pub fn enforce<T: FieldElement>(mut file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
```
I try to always do this.

Schaeff commented on parser/src/powdr.lalrpop:
```diff
@@ -152,6 +152,9 @@ Machine: Machine<T> = {
 MachineArguments: MachineArguments = {
+    "(" "_" "," "_" ")" => MachineArguments::default(),
+    "(" <latch:Identifier> "," "_" ")" => MachineArguments { latch: Some(latch), function_id: None },
+    "(" "_" "," <function_id:Identifier> ")" => MachineArguments { latch: None, function_id: Some(function_id) },
```
Read-only registers are cleaner because of this bit in `vm_to_constrained`
```rust
                          // Unconstrain read-only registers when calling `_reset`
                          ReadOnly => {
                              let not_reset =
                                  build_sub(build_number(1u64), direct_reference("instr__reset"));
                              (build_mul(not_reset.clone(), lhs), build_mul(not_reset, rhs))
                          }
```

Schaeff commented on test_data/polygon-hermez/main.pil:
```diff
@@ -74,7 +74,7 @@ namespace Main(%N);
     pol commit binOpcode;
     pol commit assert;
-    pol commit repeat, call, return;
+    pol commit repeat, call, _return;
```
#501

Schaeff commented on analysis/README.md:
```diff
@@ -0,0 +1,54 @@
+
+This process introduces instructions and registers, and returns the modified machine along with the ROM.
+
+##### ROM to fixed
```
Right. Constrained.

Schaeff commented on asm_to_pil/src/lib.rs:
```diff
@@ -1,1081 +1,129 @@
-                (location, object)
+            .map(|(name, m)| {
+                (name, {
+                    let (m, rom) = generate_machine_rom(m);
```
I don't agree. I'm trying to rely as little as possible on runtime assumptions. I agree it's a bit awkward to have a single type for both VMs and CMs but prefer relying on the type of `generate_machine_rom` (which returns an optional ROM) rather than a runtime check inside its implementation.

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,54 @@
+
+## Virtual and constrained machines
+
+| Machine               | Virtual | Constrained |
```
agree

leonardoalt commented on asm_to_pil/src/lib.rs:
```diff
@@ -1,952 +1,127 @@
-                self.negate_assignment_value(self.process_assignment_value(*expr))
+    pub fn parse_instruction<T: FieldElement>(input: &str) -> Instruction<T> {
+        match parser::powdr::InstructionDeclarationParser::new()
+            .parse(&format!("instr dummy {input}"))
```
why dummy?

Schaeff replied:
I was reusing a parser for the whole instruction to get only the right hand side (and ignoring the name). Refactored to using a parser for the right hand side only.

leonardoalt commented on asm_to_pil/src/romgen.rs:
```diff
@@ -0,0 +1,405 @@
+};
+
+/// Substitute all visited columns inside expressions of `s`
+/// This *only* applies to expressions, so for example identifiers the left hand side of statements is not substituted
```
```suggestion
/// This *only* applies to expressions, so for example identifiers in the left hand side of statements are not substituted
```

Schaeff replied:
done

leonardoalt commented on asm_to_pil/src/romgen.rs:
```diff
@@ -0,0 +1,405 @@
+_sink::
+_loop;
+// END BATCH
+"#
```
Nice!

leonardoalt commented on asm_to_pil/src/vm_to_constrained.rs:
```diff
@@ -0,0 +1,995 @@
+        let instruction_flag = format!("instr_{instruction_name}");
+        self.create_witness_fixed_pair(s.start, &instruction_flag);
+        // it's part of the lookup!
+        //self.pil.push(constrain_zero_one(&col_name));
```
remove comment?

Schaeff replied:
done

leonardoalt commented on asm_to_pil/src/vm_to_constrained.rs:
```diff
@@ -0,0 +1,995 @@
+                    Input::Literal(param.name, LiteralKind::UnsignedConstant)
+                }
+                None => Input::Register(param.name),
+                Some(ty) => panic!("param type must be nothing or label, found `{ty}`"),
```
what about signed and unsigned above?

Schaeff replied:
Removed as this should be done in the type checker

leonardoalt reviewed (commented):
Nice! I think this is almost there.

Mostly small comments below. My main one now is still regarding docs/README: I agree ideally we should simply have tests for each pass which have to be updated when code is updated, but for now I think a small e2e example in the new README will all steps would be nice.
Tomorrow I'll run tests myself and read the diff to make sure I fully understand all passes.

I'm still not a big fan of all the unreserved hardcoded keywords that may cause collision with user defined identifiers, but I think that's fine for now. I opened https://github.com/powdr-labs/powdr/issues/508 to track that.

Schaeff commented on analysis/src/block_enforcer.rs:
```diff
@@ -0,0 +1,38 @@
+        let function_id = machine.function_id.as_ref().unwrap();
+        let latch = machine.latch.as_ref().unwrap();
+        let last_step = "_block_enforcer_last_step";
+        let f_no_change = "_function_id_no_change";
```
Correct, nothing is reserved. I think one pragmatic way to fix this for now is to arbitrary disallow identifiers which start with `_` in the parser.

Schaeff replied:
All correct.

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,623 @@
+
+#### Batcher
+
+The batcher groups statements within each function into batches which can be executed in the same execution step. In our example, we gain 33% in the main machine by batching the first label with the following statement.
```
which following statement?

Schaeff replied:
Fixed, didnt have the latest version of the example in here

leonardoalt replied:
I still don't get it. The next diff is still the END BATCH comments only.

Schaeff replied:
Yes, the batcher only inserts the batch delimitations. It does not change the statements.

leonardoalt replied:
Ah, I think I misunderstood the text. So "the following statement" refers to the statement after said label, right? I imagined a statement after the sentence

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,623 @@
+-               return;
+-               // END BATCH
++
+// the body of functions is removed and they are assigned their function id. Their inputs and outputs are mapped to the relevant registers
```
```suggestion
// function bodies are removed and functions are assigned their function id. Their inputs and outputs are mapped to the relevant registers
```

leonardoalt replied:
but this hasn't been done here yet, right?

leonardoalt replied:
ah sorry, they are

leonardoalt commented on analysis/README.md:
````diff
@@ -0,0 +1,623 @@
++       function main<2> {
++
+```
+
````
Nice!

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,623 @@
+- Enable non-deterministically jumping to one of these labels by setting `function_id` to the index of the label
+
+This process introduces instructions and registers, and returns the modified machine along with the ROM.
+
```
Can you also add an example of rom gen with the dispatcher?

Schaeff replied:
Added the two roms for the example

leonardoalt replied:
nice

leonardoalt commented on analysis/README.md:
````diff
@@ -0,0 +1,623 @@
++               _function_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
++               (_function_id_no_change * (_function_id' - _function_id)) = 0;
+```
+
````
This looks a bit off. The `}` on line 265 is extra and the `{` online 269 isn't closed. Also these are identical, as explained above. But how does the compiler treat this?

Schaeff replied:
I adjusted the display for the diff to be clearer. I don't understand the last question?

leonardoalt replied:
I think I get it now. Each of these identical blocks are part of different machines, so there is no conflict. Maybe it makes sense to just show one block here (instead of 2) and just say each machine gets one of them (which is already said).

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,623 @@
+
+### Airgen
+
+Airgen takes machines which are only left with constraints and external instructions and instanciates them as a tree of AIR objects. Objects can point to each other using links, which encode the interaction between different machines.
```
```suggestion
Airgen takes machines which are only left with constraints and external instructions and instantiates them as a tree of AIR objects. Objects can point to each other using links, which encode the interaction between different machines.
```

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,623 @@
+                (_function_id_no_change * (_function_id' - _function_id)) = 0;
+        }
+
+        instr identity X -> Y { sub.identity }
```
This syntax looks kinda weird, as if `sub.identity` needs to be a true constraint. What happened to the previous `instr identity X -> Y = sub.identity;`?

Schaeff replied:
This was an issue in Display. Fixed.

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,623 @@
+
+### Linker
+
+The linker takes a tree of machines and instanciates specific structures to make proofs about. In our current implementation, it simply collates all AIR objects next to each other in a single table.
```
```suggestion
The linker takes a tree of machines and instantiates specific structures that will be used in proof generation. In our current implementation, it simply collates all AIR objects next to each other in a single table.
```

leonardoalt commented on analysis/README.md:
```diff
@@ -0,0 +1,623 @@
+_function_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
+(_function_id_no_change * (_function_id' - _function_id)) = 0;
+// Links:
+// // // instr identity with params  X -> Y links to // // function identity with id 2 with params  _input_0 -> _output_0 in object at location main_sub with latch "instr_return" and function_id "_function_id"
```
Not too important but redundant `//` here and later in the same line

leonardoalt replied:
also, where are these links?

Schaeff replied:
Fixed. What do you mean where? They belong to the object.

leonardoalt replied:
yes but at first I thought the comments were referring to something right underneath or something close

georgwiese commented:
🎉
